### PR TITLE
Make executeCheckAndRecoverFunction show unhandled analysis entries.

### PR DESCRIPTION
**Pull request by sjmudd**
Current code does not make this very visible and the code path should not get here anyway. However, if it does it indicates that the analysis phase is not completely correct or we are not handling the situation correctly. Both situations probably need to be identified and resolved.

- [x] contributed code is using same conventions as original code
- [x] code is formatted via `gofmt` (please avoid `goimports`)
- [x] code is built via `./build.sh`
- [x] code is tested via `go test ./go/...`

### CODE REVIEW

File: go/logic/topology_recovery.go [hunk 1]
```diff
@@ -1201,14 +1201,15 @@ func executeCheckAndRecoverFunction(analysisEntry inst.ReplicationAnalysis, cand
...
-	if checkAndRecoverFunction == nil {
+		// Right now this is mostly causing noise with no clear action.
+		// Will revisit this in the future.
+		// case inst.AllMasterSlavesStale:
+		// 	checkAndRecoverFunction = checkAndRecoverGenericProblem
+	default:
 		// Unhandled problem type
+		log.Warningf("executeCheckAndRecoverFunction: Ignoring unhandled analysisEntry: %+v; key: %+v",
+			analysisEntry.Analysis, analysisEntry.AnalyzedInstanceKey)
+
```
Comment (shlomi-noach) [inline go/logic/topology_recovery.go]: Incorrect change of logic. Consider the behavior of `case inst.UnreachableMaster:`: it doesn't assign a `checkAndRecoverFunction` but instead issues an immediate `go emergentlyReadTopologyInstanceReplicas(...)`.

It should then `return false, nil, nil`. But in your code change it _does not_ return, but instead proceeds to execute the recovery function. Such function does not exist and proceeding is wrong.

**Review by shlomi-noach — CHANGES_REQUESTED** [review]

File: go/logic/topology_recovery.go (hunk 1 above)
Comment (sjmudd) [reply go/logic/topology_recovery.go]: Ok. Will handle this properly shortly. Thanks for the heads up.

**Comment by sjmudd** [comment]
Hi Shlomi. thanks for brining up the problem with the change I proposed. it does look as if I'd missed the meaning of what was going on here. Bottom line, the original code has a reference to unhandled exceptions and I wanted to see if there were any and why.  I can change the code therefore to do the check properly by using the 

```
if checkAndRecoverFunction == nil
```

and adding a check to log if `if analysisEntry.Analysis != inst.NoProblem`.

Having said that I hadn't realised but in the same function `executeCheckAndRecoverFunction` you call in some cases `go emergentlyXXXX` which seems to be designed to trigger an instant discovery of one or more servers, but currently that's handled outside of the concurrency queueing mechanism.  For discovering a single server this may not be an issue but `emergentlyReadTopologyInstanceReplicas` bypasses the discovery queue which is intended to limit the number of concurrent concurrency attempts.  For a large replication chain you may trigger an extra X number of discoveries which might not be good, and might perhaps trigger later writes to the orchestrator backend to hit `max_connections`.

If you agree that analysis is correct then I'd like to modify these functions to "queue" their discoveries, but rather than doing this "at the end of the queue" to push the discovery request to the beginning so they get serviced first. That seems like a reasonable compromise etc.

This change in behaviour would seem to be fine, unless `executeCheckAndRecoverFunction` repeatedly triggers these `emergentlyXXX` functions, something which seems to be something I'm seeing as basically it bypasses the current FIFO discovery queue.

So maybe this is something to consider?

**Comment by sjmudd** [comment]
I've pushed a further change to log only if the analysis is not `NoProblem` and is not "handled" which I think should resolve your concerns.

**Comment by sjmudd** [comment]
The current patch, if you pull the latest changes in the branch, will only log that something has been missed if the problem is not NoProblem and if `if checkAndRecoverFunction == nil` is called. So no changes in behaviour: just some extra logging which hopefully will be useful. I can then follow up on that if I see (as I am) repeated occurrences of this being triggered.

So would you accept the PR as it is now so we can make progress?

Thanks for your thoughts on this.

File: go/logic/topology_recovery.go [hunk 2]
```diff
@@ -1209,6 +1209,11 @@ func executeCheckAndRecoverFunction(analysisEntry inst.ReplicationAnalysis, cand
 
 	if checkAndRecoverFunction == nil {
 		// Unhandled problem type
+		if analysisEntry.Analysis != inst.NoProblem {
+			log.Warningf("executeCheckAndRecoverFunction: Ignoring unhandled analysisEntry: %+v; key: %+v",
+				analysisEntry.Analysis, analysisEntry.AnalyzedInstanceKey)
+		}
+
```
Comment (shlomi-noach) [inline go/logic/topology_recovery.go:1216]: :+1:

**Review by shlomi-noach — APPROVED** [review]

**Pull request merged**